crates/win-api-wrappers/src/wts.rs

```diff
@@ -87,24 +98,232 @@ impl Drop for WTSVirtualChannel {
     }
 }
 
+#[repr(i32)]
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum WTSConnectState {
+    Active = windows::Win32::System::RemoteDesktop::WTSActive.0,
+    Connected = windows::Win32::System::RemoteDesktop::WTSConnected.0,
+    ConnectQuery = windows::Win32::System::RemoteDesktop::WTSConnectQuery.0,
+    Shadow = windows::Win32::System::RemoteDesktop::WTSShadow.0,
+    Disconnected = windows::Win32::System::RemoteDesktop::WTSDisconnected.0,
+    Idle = windows::Win32::System::RemoteDesktop::WTSIdle.0,
+    Listen = windows::Win32::System::RemoteDesktop::WTSListen.0,
+    Reset = windows::Win32::System::RemoteDesktop::WTSReset.0,
+    Down = windows::Win32::System::RemoteDesktop::WTSDown.0,
+    Init = windows::Win32::System::RemoteDesktop::WTSInit.0,
+}
+
+impl TryFrom<WTS_CONNECTSTATE_CLASS> for WTSConnectState {
+    type Error = anyhow::Error;
+
+    fn try_from(v: WTS_CONNECTSTATE_CLASS) -> Result<Self, Self::Error> {
+        match v {
+            windows::Win32::System::RemoteDesktop::WTSActive => Ok(WTSConnectState::Active),
+            windows::Win32::System::RemoteDesktop::WTSConnected => Ok(WTSConnectState::Connected),
+            windows::Win32::System::RemoteDesktop::WTSConnectQuery => Ok(WTSConnectState::ConnectQuery),
+            windows::Win32::System::RemoteDesktop::WTSShadow => Ok(WTSConnectState::Shadow),
+            windows::Win32::System::RemoteDesktop::WTSDisconnected => Ok(WTSConnectState::Disconnected),
+            windows::Win32::System::RemoteDesktop::WTSIdle => Ok(WTSConnectState::Idle),
+            windows::Win32::System::RemoteDesktop::WTSListen => Ok(WTSConnectState::Listen),
+            windows::Win32::System::RemoteDesktop::WTSReset => Ok(WTSConnectState::Reset),
+            windows::Win32::System::RemoteDesktop::WTSDown => Ok(WTSConnectState::Down),
+            windows::Win32::System::RemoteDesktop::WTSInit => Ok(WTSConnectState::Init),
+            _ => Err(anyhow::anyhow!("Invalid WTS_CONNECTSTATE_CLASS: {}", v.0)),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct WTSSessionInfo {
+    pub session_id: u32,
+    pub win_station_name: String,
+    pub state: WTSConnectState,
+}
+
+impl WTSSessionInfo {
+    /// # Safety
+    ///
+    /// - The `pWinStationName` pointer needs to be valid for reads up until and including the next `\0`.
+    unsafe fn new(wts_session: &WTS_SESSION_INFOW) -> anyhow::Result<Self> {
+        Ok(Self {
+            session_id: wts_session.SessionId,
+            win_station_name: unsafe { wts_session.pWinStationName.to_string()? },
```

Comment:
Question @CBenoit: I understand why the linter wants me to add the `unsafe` block here; but is it necessary to provide a SAFETY comment also - given that it's the same precondition documented at the function level?

Reply:
Yes, absolutely. The rationale is that someone reading this code has no idea that the preconditions are the same, will be confused, and end up looking up the documentation again to double check. (Personal experience. :joy: )
It’s probably not too bad when skimming in a code editor with a rich LSP integration because you can hover over the function to read the documentation. But you don’t always have that: reading code in GitHub, or even just reading the code from a Linux machine where this windows-specific code is disabled.

It’s enough to go with something like
```rust
// SAFETY: Same preconditions as the current function.
```